### PR TITLE
Fall back to join ping if not master

### LINKED ISSUE #97313: Avoid getStateForMasterService during join validation
https://github.com/elastic/elasticsearch/pull/97304 removes several places that might do an expensive rebuild of the cluster state via `getStateForMasterService`. There's one other call we should be able to remove, in `JoinValidationService`, but it's a little tricky since we rely on seeing the last-accepted state for the cluster UUID validation, and it's not obviously ok to expose the last-accepted state without the no-master block.

This is a fairly minor issue because we do not do this join validation during elections, only if we're the elected master, and in that case `getStateForMasterService` is cheap. But that's not guaranteed, and we should strengthen this to avoid the that expensive step whenever possible.

### PR DESCRIPTION
**Pull request by DaveCTurner**
The `JoinValidationService` doesn't need to send the cluster state if
we're not currently the master - instead it can just send a join ping.
This means it doesn't need to do any work adjusting the blocks.

Closes #97313

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java [hunk 1]
```diff
@@ -1478,6 +1469,20 @@ ClusterState getStateForMasterService() {
...
     }
 
+    private ClusterState getStateForJoinValidationService() {
+        synchronized (mutex) {
+            // similar to getStateForMasterService, but do not rebuild the state if not currently the master
+            final ClusterState clusterState = coordinationState.get().getLastAcceptedState();
+            assert clusterState.nodes().getLocalNode() != null;
+            if (mode != Mode.LEADER
+                || clusterState.term() != getCurrentTerm()
+                || clusterState.nodes().isLocalNodeElectedMaster() == false) {
```
Comment (kingherc) [inline server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java:1479]: > `|| clusterState.nodes().isLocalNodeElectedMaster() == false) {`

This line is new and previously it was an assertion. Why is that?

**Review by kingherc — COMMENTED** [review]
Only a couple of comments. I do not have much expertise here, so I think I'd let @fcofdez comment & approve if it's not urgent.

File: server/src/main/java/org/elasticsearch/cluster/coordination/JoinValidationService.java [hunk 2]
```diff
@@ -153,26 +155,39 @@ public void validateJoin(DiscoveryNode discoveryNode, ActionListener<TransportRe
...
-                                "failure when sending a join validation request from [%s] to [%s]",
-                                transportService.getLocalNode().descriptionWithoutAttributes(),
-                                discoveryNode.descriptionWithoutAttributes()
-                            ),
-                            e
-                        )
-                    );
-                }), i -> TransportResponse.Empty.INSTANCE, ThreadPool.Names.CLUSTER_COORDINATION)
-            );
+            final var responseHandler = new ActionListenerResponseHandler<>(listener.delegateResponse((l, e) -> {
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/coordination/JoinValidationService.java:158]: This branch needs to be adjusted to handle the case where `clusterStateSupplier` returns `null`. I'd rather do that than have a different `clusterStateSupplier` for the BwC case.

File: server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java:1479]: I don't see where this used to be an assertion, sorry if I'm missing something. This is really just folding in the `isLocalNodeElectedMaster` check that previously was here:

https://github.com/elastic/elasticsearch/blob/0b5f19bdfcd639134d36674277e748d72d4926d2/server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java#L695

File: server/src/main/java/org/elasticsearch/cluster/coordination/JoinValidationService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kingherc) [reply server/src/main/java/org/elasticsearch/cluster/coordination/JoinValidationService.java:158]: Sounds good thanks for the explanation!

**Review by fcofdez — APPROVED** [review]
LGTM

**Pull request merged**